[Controls] Port Modifier Buttons and Speed Modifiers from SoH.

Glought opened the pull request:
Port Modifier Buttons and Speed Modifiers from SoH.

* Adds "M1" and "M2" buttons extra buttons like the y and x buttons on the GameCube Controller, or a keyboard key can be assigned.

Port Speed Modifiers from SoH. -> #939

When Walk Speed Modifier is enabled Link's walk speed will be increased depending on the "Walk Modifier 1 or Modifier 2" setting. 
"Walk Modifier 1" is "M1".
"Walk Modifier 2" is "M2".
Default:100%
Max:1500%
Min:0%

When Swim Speed Modifier is enabled Link's swim speed will be increased depending on the "Swim Modifier 1 or Modifier 2" setting. 
"Swim Modifier 1 is M1.
"Swim Modifier 2 is M2.
Default:100%
Max:875%
Min:0%

"Toggle modifier instead of holding" setting will keep the Modifier active till the modifier button is pressed again.
Modifier button toggles are reset when 2S2H is reset.

<img width="1720" height="899" alt="2ship Controls Modifier Buttons" src="https://github.com/user-attachments/assets/c79a6a36-079b-40b1-b6a3-4bc32157e445" />


<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4878731300.zip)
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4878788493.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4879084234.zip)
<!--- section:artifacts:end -->

Glought committed:
[Controls] Port Modifier Buttons from SoH.
* Adds "M1" and "M2" buttons extra buttons like the y and x buttons on the
  GameCube Controller, or a keyboard key can be assigned.

Port Speed Modifiers from SoH.

When Walk Speed Modifier is enabled Link's walk speed will be increased
depending on the "Walk Modifier 1 or Modifier 2" setting.
"Walk Modifier 1" is "M1".
"Walk Modifier 2" is "M2".
Default:100%
Max:1500%
Min:0%

When Swim Speed Modifier is enabled Link's swim speed will be increased
depending on the "Swim Modifier 1 or Modifier 2" setting.
"Swim Modifier 1 is M1.
"Swim Modifier 2 is M2.
Default:100%
Max:875%
Min:0%

"Toggle modifier instead of holding" setting will keep the Modifier
active till the modifier button is pressed again.
Modifier button toggles are reset when 2S2H is reset.

Glought committed:
Ran Clang Format

Glought committed:
Renamed "initWalkSpeedFunc" to "initFunc" in "LinkSpeedModifier.cpp"

mckinlee commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,89 @@
+    });
+
+    COND_VB_SHOULD(VB_SPEED_MODIFIER_SWIM, CVAR_SWIM && CVAR_SPEED, {
+        Player* player = GET_PLAYER(gPlayState);
```
This appears to be unused.

Glought replied:
oops missed that.

mckinlee commented on mm/2s2h/BenPort.h:
```diff
@@ -3,6 +3,9 @@
 #pragma once
 
+#define BTN_CUSTOM_MODIFIER1 0x0040
+#define BTN_CUSTOM_MODIFIER2 0x0080
```
Could this conflict with BTN_RESET in controller.h? It is also 0x0080.

Glought replied:
Nothing seems to use "BTN_RESET"  just the define in controller.h is found.

Glought replied:
I changed the Custom Modifiers to avoid any issues.

Glought replied:
Turns out I can't change them because the input viewer registers 3 buttons being pressed when I press either modifier button.

mckinlee replied:
No biggie. You can revert them back. I was only asking, not suggesting you change them. Sorry for confusion lol!

Glought replied:
I wanted to avoid any potential issues if "BTN_RESET" ever gets used for something oh well.

mckinlee replied:
Yeh, oh well. If you want, you could put a comment referencing this file should someone wanna do something with it.  Great job on this, btw. I wanna spend a little more time doing in-game testing, but codewise this has my approval.

mckinlee commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,89 @@
+                speedToggle2 = !speedToggle2;
+            }
+        }
+    });
```
This is fine as is, but I do have a suggestion. You could use OnPassPlayerInputs instead. That hook provides the processed input that the player update uses.

mckinlee commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+static RegisterShipInitFunc initFunc(RegisterLinkSpeedModifier,
+                                     { CVAR_SPEED_MODIFIER_NAME, CVAR_WALK_MODIFIER_NAME, CVAR_SWIM_MODIFIER_NAME });
```
Should this also be dependent on `gSettings.SpeedModifier.Toggle`?

Glought committed:
Made changes based on mckinlee's suggestions.

Glought committed:
Made it so only one Modifier button is toggled at a time.
For example if you press "M1" it will untoggle "M2"

mckinlee commented on mm/2s2h/BenGui/BenInputEditorWindow.cpp:
```diff
@@ -22,6 +27,7 @@ void BenInputEditorWindow::InitElement() {
 
     mButtonsBitmasks = { BTN_A, BTN_B, BTN_START, BTN_L, BTN_R, BTN_Z, BTN_CUP, BTN_CDOWN, BTN_CLEFT, BTN_CRIGHT };
     mDpadBitmasks = { BTN_DUP, BTN_DDOWN, BTN_DLEFT, BTN_DRIGHT };
+    mModifierButtonsBitmasks = { BTN_CUSTOM_MODIFIER1, BTN_CUSTOM_MODIFIER2 };
```
I don't see where this is used. Whats the plan for this?

Glought replied:
Honestly idk i just looked into it its not used anywhere on SoH but the same can be said for `mDpadBitmasks`  and `mButtonsBitmasks ` both are not used anywhere on SoH and 2S2H. Maybe its old code from a old version of the input editor. I'll delete my lines since i just tested it commented out and it still works.

mckinlee replied:
Sounds good. Thanks!

mckinlee commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,93 @@
+
+        *maxSpeed *= swimMod;
+
+        Math_AsymStepToF(speed, *speedTarget * 0.8f * swimMod, *incrStep, (fabsf(*speed) * 0.02f) + 0.05f);
```
So since this doesn't modify the `*should*`, `GameInteractor_Should` always returns `true`, so `Math_AsymStepToF` runs twice per frame with different parameters.

Glought committed:
Removed "mModifierButtonsBitmasks" in "BenInputEditorWindow.cpp" wasn't
needed.
Fixed issue with "COND_VB_SHOULD(VB_SPEED_MODIFIER_SWIM," forgot to set
"*should"
Ran Clang Format.

Glought committed:
Revert BTN_CUSTOM_MODIFIER1 back to 0x0040
Revert BTN_CUSTOM_MODIFIER2 back to 0x0080

Glought committed:
Moved "VB_SPEED_MODIFIER_WALK" and "VB_SPEED_MODIFIER_WALK" to
"GameInteractor_VanillaBehavior.h"

Eblo commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,94 @@
+        Math_AsymStepToF(speed, *speedTarget * 0.8f * swimMod, *incrStep, (fabsf(*speed) * 0.02f) + 0.05f);
+    });
+
+    COND_HOOK(OnPassPlayerInputs, CVAR_WALK && CVAR_SPEED || CVAR_SWIM && CVAR_SPEED, [](Input* input) {
```
What is the intended order of operations in this hook condition? I think some parentheses are due here. I imagine what you want is `(CVAR_WALK && CVAR_SPEED) || (CVAR_SWIM && CVAR_SPEED)`, but that's not clear.

Edit: And if that's the case, it could be reduced to `CVAR_SPEED && (CVAR_WALK || CVAR_SWIM)`.

Eblo commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,94 @@
+        }
+    });
+
+    COND_HOOK(OnConsoleLogoUpdate, CVAR_WALK && CVAR_SPEED || CVAR_SWIM && CVAR_SPEED, []() {
```
Ditto.

Eblo commented on mm/2s2h/GameInteractor/GameInteractor_VanillaBehavior.h:
```diff
@@ -1979,6 +1979,26 @@ typedef enum {
+    // - '*f32' (maxSpeed)
+    // - '*f32' (speed)
+    // - '*f32' (speedTarget)
+    // - 'f32'  (swimMod)
```
`swimMod` is something used in the enhancement code, but it's not an arg passed through the VB.
```suggestion
```

Also, we're trying to keep these VBs alphabetized if possible, so `VB_SPEED_MODIFIER_WALK` and `VB_SPEED_MODIFIER_SWIM` should be higher in the list instead of appended to the end.

Eblo commented on mm/2s2h/Enhancements/Player/LinkSpeedModifier.cpp:
```diff
@@ -0,0 +1,94 @@
+    });
+
+    COND_HOOK(OnConsoleLogoUpdate, CVAR_WALK && CVAR_SPEED || CVAR_SWIM && CVAR_SPEED, []() {
+        speedToggle1 = 0;
```
Any reason you're using `0` here but `false` elsewhere for these bools?

Eblo reviewed (commented):
Makes sense to me for the most part, but the errant build failure kept me from actually testing this. Some minor notes.

Glought committed:
Made changes suggested by Eblo.

Glought commented:
I made the suggested changes

Eblo commented on mm/2s2h/GameInteractor/GameInteractor_VanillaBehavior.h:
```diff
@@ -1793,6 +1793,25 @@ typedef enum {
+    // - '*f32' (maxSpeed)
+    // - '*f32' (speed)
+    // - '*f32' (speedTarget)
+    VB_SPEED_MODIFIER_SWIM,
```
Swap these so they're alphabetical, and we're golden.

```suggestion
    // #### `result`
    // ```c
    // true
    // ```
    // #### `args`
    // - '*f32' (incrStep)
    // - '*f32' (maxSpeed)
    // - '*f32' (speed)
    // - '*f32' (speedTarget)
    VB_SPEED_MODIFIER_SWIM,

    // #### `result`
    // ```c
    // true
    // ```
    // #### `args`
    // - None
    VB_SPEED_MODIFIER_WALK,
```
